### PR TITLE
Revisit groovy delegates and property/method lookup

### LINKED ISSUE #276: Env. and Params. values not accessible without proper prefix
In actual jenkins pipeline params and Env are accessible without their respective prefixes. This does not work in PipelineUnit, causing scripts that work on Jenkins to fail on unit test

### PR DESCRIPTION
**Pull request by Willem1987**
Trying to leverage the default groovy delegating system to find properties and methods.

Trying to fix #276

### CODE REVIEW

**Comment by nre-ableton** [comment]
Please remove all indentation changes in this branch, as they make this PR harder to review. I agree that there is some inconsistency in the codebase, but we need a better approach to solving it. This is already on the radar, please see #237.

**Comment by Willem1987** [comment]
I think i removed the whitespace changes

**Comment by Willem1987** [comment]
Should also fix #280 with latest commit

**Comment by Willem1987** [comment]
Sorry for all the commits. I was having issues with shelved changes and different changelists with reverting and such. Tests all ran successfully locally.

I saw whitespace changes still in ParametersDeclaration which i can remove if needed.

**Comment by nre-ableton** [comment]
Ok, it's looking a bit better, but the branch history still needs some work. 987af5caf669ddd76ac2bae637033536c7d238d2 should be squashed to a658ae4b8661a0cf14ce8860b2544fd1dfbd1548, and c74addda8f02003c1613b07dab874d09e1644e30 appears to have a default commit message created by your IDE (or some other tool).

**Comment by Willem1987** [comment]
How would i go about this? c74addd Should have been part of Remove owner param in Create component. I'm trying with intelliJ as well as TortoiseGit but i don't usually update branch history

**Comment by Willem1987** [comment]
It is now 1 commit

**Comment by nre-ableton** [comment]
@Willem1987 Great, thanks. I'll try to take a look at this PR (and the others) today or tomorrow.

**Review by nre-ableton — COMMENTED** [review]
For now, I just have a few superficial suggestions. I'd like to check out this branch and play around with it a bit in order to really examine the new implementation.

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AnyOfDeclaration.groovy [hunk 1]
```diff
@@ -26,7 +26,7 @@ class AnyOfDeclaration extends WhenDeclaration {
         return false
     }
 
-    boolean execute(Object delegate) {
+    Boolean execute(Object delegate) {
```
[the remark this answers is not in the document]
Comment (Willem1987) [reply src/main/groovy/com/lesfurets/jenkins/unit/declarative/AnyOfDeclaration.groovy:29]: AnyOfDeclaration.groovy: 29: The return type of java.lang.Boolean execute(java.lang.Object) in com.lesfurets.jenkins.unit.declarative.AnyOfDeclaration is incompatible with boolean in com.lesfurets.jenkins.unit.declarative.WhenDeclaration
. At [29:5]  @ line 29, column 5.
       Boolean execute(Object delegate) {

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/DeclarativePipeline.groovy [hunk 2]
```diff
@@ -16,15 +16,11 @@ class DeclarativePipeline extends GenericPipelineDeclaration {
         properties.put('scm', 'scm')
     }
 
-    def getParams() {
```
[the remark this answers is not in the document]
Comment (Willem1987) [reply src/main/groovy/com/lesfurets/jenkins/unit/declarative/DeclarativePipeline.groovy:19]: params is a property/variable available on binding. 

Params_Jenkinsfile still works.

Depending on if params is available and XYZ in new situation i can imagine it throwing a PropertyMissingException, but im not sure returing null would help the pipeline there

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/NotDeclaration.groovy [hunk 3]
```diff
@@ -2,7 +2,7 @@ package com.lesfurets.jenkins.unit.declarative
 
 class NotDeclaration extends WhenDeclaration {
 
-    boolean execute(Object delegate) {
+    Boolean execute(Object delegate) {
```
[the remark this answers is not in the document]
Comment (Willem1987) [reply src/main/groovy/com/lesfurets/jenkins/unit/declarative/NotDeclaration.groovy:5]: Same issue as AnyOfDeclaration. It overrides the def execute in GenericPipelineDeclaration which seems to infer Object. boolean does not align with that.

**Comment by Willem1987** [comment]
As far as i can see the repo is public so you should be able to checkout. All the tests that were there still work and i updated Agent_bindings_Jenkinsfile to include a method in the script which now works.

Only thing that does not work is accessing the AGENT var from within the agentName() method, but this might be a groovy scoping thing.

**Review by nre-ableton — APPROVED** [review]
So this PR initially looks ok to me, but admittedly it touches some things here I have no experience with (such as declarative pipelines). That said, the tests still pass, and we certainly have pipelines which use `env.XYZ` and `params.XYZ` syntax, so it's probably not harmful.

I would really appreciate if another admin could have a look before merging, though. Perhaps @stchar? :pray:

**Comment by stchar** [comment]
Hi @Willem1987, @nre-ableton, 
This PR is huge I hope can review it today

**Review by stchar — APPROVED** [review]
Looks good to me

**Comment by stchar** [comment]
Merged. Thank you for all your efforts!

**Comment by Willem1987** [comment]
Is there any upcoming release? I think with the pull requests merged i should be able to use the new version in our project.

**Comment by stchar** [comment]
I've just releases new version, usually it takes 1-2days to replicate artifacts to public maven repository

**Comment by stchar** [comment]
you can always checkout the tag (v1.8) and run `./gradlew publishToMavenLocal `

**Pull request merged**